File: lib/mydia/p2p/relay_list.ex
defmodule Mydia.P2p.RelayList do
  @moduledoc """
  Resolves which iroh relays this install should use.

  The relay hostname used to be compiled in, so moving it meant shipping a new
  server image and waiting for every install to upgrade. This reads the list
  from the metadata relay's `/client-config` at boot instead, so a relay move
  is a relay deploy.

  Precedence, highest first:

    1. An explicit `IROH_RELAY_URL`. Used alone, and the fetch is skipped
       entirely. Someone who set it did so to pin their install to one relay,
       so merging mydia's list into theirs would defeat the setting.
    2. `p2p.relays` from a successful fetch.
    3. The last successful fetch, read from disk.
    4. The compiled-in default.

  iroh's own public relays are appended underneath whatever this returns, by
  `build_relay_mode` in `native/mydia_p2p_core/src/lib.rs`. That is why losing
  a mydia relay degrades rather than breaks, and why this module never needs to
  carry them.

  The fetch goes to the *configured* metadata relay, not a hardcoded
  `relay.mydia.dev`, so a self-hoster running their own relay serves their own
  list and no install gains a network dependency it did not already have.

  `player/lib/core/p2p/relay_list.dart` is the Dart twin of this module and
  implements the same four levels and the same validation.
  """

  require Logger

  alias Mydia.Metadata

  @default_relay_url "https://cae1-1.relay.mydia.dev"
  @timeout_ms 3_000
  @path "/client-config"

  @type source :: :override | :fetched | :cached | :default

  @doc """
  The relay compiled into this build, used when nothing else is available.
  """
  @spec default_relay_url() :: String.t()
  def default_relay_url, do: @default_relay_url

  @doc """
  Resolves the relay list.

  Returns the URLs and which precedence level produced them, so the caller can
  log something honest at boot.

  Options exist for tests: `:cache_path`, `:base_url`, `:req_options` and
  `:override`. In production all four come from the environment.
  """
  @spec resolve(keyword()) :: {[String.t()], source()}
  def resolve(opts \\ []) do
    case override(opts) do
      [_ | _] = urls -> {urls, :override}
      [] -> resolve_from_relay(opts)
    end
  end

  defp resolve_from_relay(opts) do
    cache_path = Keyword.get(opts, :cache_path) || default_cache_path()

    case fetch(opts) do
      [_ | _] = urls ->
        write_cache(cache_path, urls)
        {urls, :fetched}

      [] ->
        case read_cache(cache_path) do
          [_ | _] = urls -> {urls, :cached}
          [] -> {[@default_relay_url], :default}
        end
    end
  end

  defp override(opts) do
    raw =
      case Keyword.fetch(opts, :override) do
        {:ok, value} -> value
        :error -> System.get_env("IROH_RELAY_URL")
      end

    # System.get_env/1 returns "" for a variable explicitly set empty, and ""
    # is truthy in Elixir, so trimming is what actually treats blank as unset.
    case raw do
      nil -> []
      value -> value |> String.trim() |> List.wrap() |> valid_urls()
    end
  end

  defp fetch(opts) do
    base_url = Keyword.get(opts, :base_url) || Metadata.default_relay_config().base_url
    req_options = Keyword.get(opts, :req_options, [])

    req =
      Req.new(
        base_url: base_url,
        receive_timeout: @timeout_ms,
        connect_options: [timeout: @timeout_ms],
        retry: false,
        headers: [accept: "application/json"]
      )
      |> Req.merge(req_options)

    case Req.get(req, url: @path) do
      {:ok, %Req.Response{status: 200, body: body}} ->
        relays_from(body)

      {:ok, %Req.Response{status: status}} ->
        Logger.warning("Relay list fetch returned HTTP #{status}, falling back")
        []

      {:error, reason} ->
        Logger.warning("Relay list fetch failed: #{inspect(reason)}, falling back")
        []
    end
  rescue
    e ->
      Logger.warning("Relay list fetch raised: #{inspect(e)}, falling back")
      []
  end

  # Unknown keys are ignored on purpose: the document is meant to grow without
  # a client change.
  defp relays_from(%{"p2p" => %{"relays" => relays}}) when is_list(relays),
    do: valid_urls(relays)

  defp relays_from(body) when is_binary(body) do
    case Jason.decode(body) do
      {:ok, decoded} -> relays_from(decoded)
      {:error, _} -> []
    end
  end

  defp relays_from(_), do: []

  defp valid_urls(urls) do
    Enum.filter(urls, fn
      url when is_binary(url) ->
        case URI.parse(url) do
          %URI{scheme: "https", host: host} when is_binary(host) and host != "" -> true
          _ -> false
        end

      _ ->
        false
    end)
  end

  defp read_cache(path) do
    with {:ok, contents} <- File.read(path),
         {:ok, urls} when is_list(urls) <- Jason.decode(contents) do
      valid_urls(urls)
    else
      _ -> []
    end
  end

  defp write_cache(path, urls) do
    path |> Path.dirname() |> File.mkdir_p!()
    File.write!(path, Jason.encode!(urls))
  rescue
    e ->
      # A cache we cannot write costs us the next offline boot, nothing more.
      Logger.warning("Could not cache the relay list at #{path}: #{inspect(e)}")
      :ok
  end

  defp default_cache_path do
    keypair_path =
      Application.get_env(:mydia, :p2p_keypair_path) ||
        raise "p2p_keypair_path is not configured"

    Path.join(Path.dirname(keypair_path), "relays.json")
  end
end

File: player/lib/core/p2p/relay_list.dart
import 'dart:async';
import 'dart:convert';

import 'package:flutter/foundation.dart' show debugPrint;
import 'package:http/http.dart' as http;

import '../auth/auth_storage.dart';
import '../relay/relay_api_client.dart' show metadataRelayBaseUrl;

/// The iroh relay compiled into this build, used when nothing else is
/// available. Overridable at build time with `--dart-define IROH_RELAY_URL`.
const String defaultIrohRelayUrl = 'https://cae1-1.relay.mydia.dev';

/// The build-time iroh relay override. Empty when not supplied.
const String buildTimeIrohRelayUrl = String.fromEnvironment('IROH_RELAY_URL');

/// Where the last successful fetch is cached.
///
/// AuthStorage rather than a file so the web build works unchanged: a browser
/// has no filesystem, and AuthStorage already abstracts that split.
///
/// Not to be confused with `_ConnectionStorageKeys.relayUrl` in
/// `core/connection/connection_provider.dart`, which records the relay a
/// paired server was reachable through. That key is a different thing and is
/// left alone.
const String relayListStorageKey = 'p2p_relays';

const Duration _timeout = Duration(seconds: 3);
const String _path = '/client-config';

/// Which precedence level produced a relay list.
enum RelayListSource { override, fetched, cached, defaultBuiltIn }

/// A resolved relay list and where it came from.
class RelayListResult {
  const RelayListResult(this.urls, this.source);

  final List<String> urls;
  final RelayListSource source;
}

/// Resolves which iroh relays this player should use.
///
/// Precedence, highest first: an explicit override (used alone, no request is
/// made), a successful fetch of the metadata relay's `/client-config`, the last
/// successful fetch from storage, then the compiled-in default.
///
/// iroh's own public relays are appended underneath whatever this returns, by
/// `build_relay_mode` in `native/mydia_p2p_core/src/lib.rs`.
///
/// `lib/mydia/p2p/relay_list.ex` is the Elixir twin of this function and
/// implements the same four levels and the same validation.
Future<RelayListResult> resolveRelayList({
  String? override,
  String baseUrl = metadataRelayBaseUrl,
  http.Client? client,
  AuthStorage? storage,
}) async {
  final effectiveOverride = _validUrls([
    if (override != null)
      override
    else if (buildTimeIrohRelayUrl.isNotEmpty)
      buildTimeIrohRelayUrl,
  ]);

  if (effectiveOverride.isNotEmpty) {
    return RelayListResult(effectiveOverride, RelayListSource.override);
  }

  final store = storage ?? getAuthStorage();
  final fetched = await _fetch(baseUrl, client);

  if (fetched.isNotEmpty) {
    await _writeCache(store, fetched);
    return RelayListResult(fetched, RelayListSource.fetched);
  }

  final cached = await _readCache(store);
  if (cached.isNotEmpty) {
    return RelayListResult(cached, RelayListSource.cached);
  }

  return const RelayListResult(
    [defaultIrohRelayUrl],
    RelayListSource.defaultBuiltIn,
  );
}

Future<List<String>> _fetch(String baseUrl, http.Client? client) async {
  final own = client == null;
  final httpClient = client ?? http.Client();

  try {
    final response = await httpClient.get(Uri.parse('$baseUrl$_path'),
        headers: {'accept': 'application/json'}).timeout(_timeout);

    if (response.statusCode != 200) {
      debugPrint(
          '[P2P] Relay list fetch returned ${response.statusCode}, falling back');
      return const [];
    }

    final decoded = jsonDecode(response.body);
    if (decoded is! Map<String, dynamic>) return const [];

    // Unknown keys are ignored on purpose: the document is meant to grow
    // without a client change.
    final p2p = decoded['p2p'];
    if (p2p is! Map<String, dynamic>) return const [];

    final relays = p2p['relays'];
    if (relays is! List) return const [];

    return _validUrls(relays);
  } catch (e) {
    debugPrint('[P2P] Relay list fetch failed: $e, falling back');
    return const [];
  } finally {
    if (own) httpClient.close();
  }
}

List<String> _validUrls(List<dynamic> urls) {
  return urls
      .whereType<String>()
      .where((url) {
        final parsed = Uri.tryParse(url.trim());
        return parsed != null &&
            parsed.scheme == 'https' &&
            parsed.host.isNotEmpty;
      })
      .map((url) => url.trim())
      .toList();
}

Future<List<String>> _readCache(AuthStorage storage) async {
  try {
    final raw = await storage.read(relayListStorageKey);
    if (raw == null) return const [];

    final decoded = jsonDecode(raw);
    if (decoded is! List) return const [];

    return _validUrls(decoded);
  } catch (e) {
    debugPrint('[P2P] Could not read the cached relay list: $e');
    return const [];
  }
}

Future<void> _writeCache(AuthStorage storage, List<String> urls) async {
  try {
    await storage.write(relayListStorageKey, jsonEncode(urls));
  } catch (e) {
    // A cache we cannot write costs us the next offline start, nothing more.
    debugPrint('[P2P] Could not cache the relay list: $e');
  }
}
